llvm/test/CodeGen/AMDGPU/mfma-no-register-aliasing.ll

```diff
@@ -636,168 +604,208 @@ define amdgpu_kernel void @test_mfma_f32_16x16x1f32(ptr addrspace(1) %arg) #0 {
-; GREEDY90A-NEXT:    global_store_dwordx4 v2, a[0:3], s[16:17]
+; GREEDY90A-NEXT:    s_nop 1
+; GREEDY90A-NEXT:    global_store_dwordx4 v0, a[12:15], s[16:17] offset:48
+; GREEDY90A-NEXT:    global_store_dwordx4 v0, a[8:11], s[16:17] offset:32
+; GREEDY90A-NEXT:    global_store_dwordx4 v0, a[4:7], s[16:17] offset:16
+; GREEDY90A-NEXT:    global_store_dwordx4 v0, a[0:3], s[16:17]
 ; GREEDY90A-NEXT:    s_endpgm
 ;
 ; GREEDY942-LABEL: test_mfma_f32_16x16x1f32:
 ; GREEDY942:       ; %bb.0: ; %bb
 ; GREEDY942-NEXT:    s_load_dwordx2 s[16:17], s[4:5], 0x24
 ; GREEDY942-NEXT:    v_mov_b32_e32 v0, 1.0
 ; GREEDY942-NEXT:    v_mov_b32_e32 v1, 2.0
-; GREEDY942-NEXT:    v_mov_b32_e32 v2, 0
 ; GREEDY942-NEXT:    s_waitcnt lgkmcnt(0)
 ; GREEDY942-NEXT:    s_load_dwordx16 s[0:15], s[16:17], 0x0
 ; GREEDY942-NEXT:    s_waitcnt lgkmcnt(0)
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a33, s15
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a32, s14
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a31, s13
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a30, s12
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a29, s11
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a28, s10
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a27, s9
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a26, s8
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a25, s7
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a24, s6
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a23, s5
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a22, s4
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a21, s3
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a20, s2
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a19, s1
-; GREEDY942-NEXT:    v_accvgpr_write_b32 a18, s0
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a0, s0
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a1, s1
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a2, s2
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a3, s3
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a4, s4
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a5, s5
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a6, s6
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a7, s7
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a8, s8
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a9, s9
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a10, s10
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a11, s11
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a12, s12
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a13, s13
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a14, s14
+; GREEDY942-NEXT:    v_accvgpr_write_b32 a15, s15
 ; GREEDY942-NEXT:    s_nop 1
-; GREEDY942-NEXT:    v_mfma_f32_16x16x1_4b_f32 a[18:33], v0, v1, a[18:33]
-; GREEDY942-NEXT:    v_mfma_f32_16x16x1_4b_f32 a[2:17], v0, v1, a[18:33]
+; GREEDY942-NEXT:    v_mfma_f32_16x16x1_4b_f32 a[0:15], v0, v1, a[0:15]
+; GREEDY942-NEXT:    v_mfma_f32_16x16x1_4b_f32 a[16:31], v0, v1, a[0:15]
 ; GREEDY942-NEXT:    s_nop 7
-; GREEDY942-NEXT:    s_nop 0
-; GREEDY942-NEXT:    v_accvgpr_mov_b32 a0, a18
-; GREEDY942-NEXT:    v_accvgpr_mov_b32 a1, a19
+; GREEDY942-NEXT:    s_nop 1
+; GREEDY942-NEXT:    v_accvgpr_mov_b32 a2, a16
```

Comment:
I looked at this in detail and I think it makes the most sense to live with it.

RegisterCoalescing actually reduces the number of copies and we end up with less registers needed. However, before this PR, the extra copies caused different register assignment, which enabled more copy propagation.

In order to resolve this, I think we would need to have RA heuristics work to facilitate more copy propagation.